### PR TITLE
Dynamax

### PR DESCRIPTION
**Pull request by AgustinGDLV**
## Description ##
This PR adds Dynamax to the expansion! The player will be able to Dynamax a Pokémon during Raid Battles (not yet implemented) or any battle while `B_FLAG_DYNAMAX_BATTLE` is active _and_ the player has a Dynamax Band in their bag. This can be used once per battle, lasts 3 turns, transforms all moves into Max Moves, and increases HP by 1.5x while active. The opponent is able to do so as well, and will always Dynamax its final remaining Pokémon.

## What is done? ##
- [x] Battle UI
  - [x] There is a toggleable "use" sprite for choosing to Dynamax.
  - [x] When Dynamax is selected, Max Moves are displayed instead of base moves, including G-Max Moves if applicable.
  - [x] There is an indicator sprite in a Dynamaxed Pokémon's healthbox.
- [x] Dynamax Features
  - [x] A battler has its HP and max HP multiplied by 1.5x when Dynamaxing.
  - [x] Dynamax expires after three turns, after which a battler has its HP and max HP divided by the above multiplier, _rounded up_.
  - [x] Dynamaxing is only possible in battle when a configurable flag is set; the player requires a Dynamax Band.
  - [x] Pokémon who can Mega Evolve or use a Z-Move cannot Dynamax.
  - [x] Every interaction listed in this [research thread](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/) and this [research video](https://www.youtube.com/watch?v=5C6JfhhXn2w) is implemented unless stated otherwise.
  - [x] The majority of the interactions above have tests written.
  - [x] TODO: Shake the Pokémon when it first Dynamaxes for cool effect!
- [x] Max Moves
  - [x] Every single Max Move and G-Max Move effect is implemented.
  - [x] Every single Max Move and G-Max Move has a test, including for some, but not all, interactions in the above research links.
  - [x] _Almost_ every single Max Move and G-Max Move have a placeholder animation.
  - [x] Max Moves change type as you'd expect (i.e Weather Ball in rain becomes Max Geyser).
  - [x] Max Moves bypass protection, except for Max Guard, except for G-Max One Blow and Rapid Flow.
  - [x]  Max Moves cannot be used against allies.
  - [x] Max Guard has a couple weird behaviors, including preventing Block but not Mean Look.
  - [x] -ate abilities should not boost the power of a Max Move.
- [x] Gigantamax Forms
  - [x] All Gigantamax Forms have proper species data and sprites.
  - [x] All Gigantamax Forms have proper cries.
  - [x] Pokémon turn into their Gigantamax Forms upon Dynamaxing.
  - [x] Pokémon revert to their base form upon switch, faint, battle end, or when Dynamax expires.
  - ~~Gigantamax form changes do not undo the effects of Soak, Magic Powder, Speed Swap, etc., as they are not "proper form changes" apparently.~~
- [x] AI
  - [x] The opponent will Dynamax if they are on their last Pokémon and are able to Dynamax.
- [x] Miscellaneous
  - [x] Leveling up does not undo the Dynamax HP multiplier or display a bugged out level-up box!
  - [x] Current HP increases with max HP when leveling up.
  - [x] Most effects that heal or damage based off a percentage of health do so using non-Dynamax HP.
 
## Give me more! ##
<details>
<summary>Not entirely exhaustive list of specific cases:</summary>

- Dynamaxed Pokémon (or Pokémon that have selected to Dynamax) can choose moves otherwise disabled by Disable, Encore, Choice Items, Gorilla Tactics, and Imprison, but **not** Taunt or Assault Vest.
- Dynamaxed Pokémon will still use Struggle if they are unable to select any moves.
- Dynamaxed Pokémon cannot be forced to switch out with phazing moves or Red Card. These moves will still deal damage if applicable and Red Card is still consumed.
- Dynamaxed Pokémon can be forced to switch out via Eject Button or Emergency Exit.
- Transformed Pokémon can Dynamax, but do not change into Gigantamax forms.
- Shedinja has 1 HP even if Dynamaxed.
- Dynamaxing causes a Pokémon's substitute to fade.
- Zacian, Zamazenta, and Eterantus cannot Dynamax, and neither can a Pokémon transformed into them.
- Gorilla Tactics and Choice Items have no effect during Dynamax.
- Destiny Bond cannot take down a Dynamax Pokémon, but Grudge still eats all of a Max Move's PP if it succeeds.
- Perish Song can take down a Dynamax Pokémon.
- Disable works against a Dynamax Pokémon if its last used move was not a Max Move, but fails otherwise.
- Mentioned previously, but any effect that damages or heals based off a proportion of HP does so based on Dynamax HP, including Pain Split, Endeavor, Leftovers, Life Orb, so on. The only exception is G-Max Finale.
- Entrainment, Skill Swap, and Wandering Spirit fail against a Dynamaxed Pokémon.
- Gastro Acid, Simple Beam, and Worry seed work against a Dynamaxed Pokémon.
- Type-changing moves appear as Max Strike (or whatever they would expect to be) in move selection, but change type on use.
- Sheer Force does not have any interactions with Max Move effects.
- Feint can hit through Max Guard, but does not break protection.
- Encore can be used to encore the base move of the last Max Move after Dynamax expires.
- Max Moves have an effect even if they hit a Substitute.
- Copycat copies the last Max Move's base move.
- Max Guard blocks some moves that bypass protection, including Block, but not Mean Look.
- G-Max Replenish recycles both or neither ally Pokémon's berries. 
- Serene Grace does not apply to G-Max Move effects.
- G-Max Chi Strike is stackable.
- G-Max Centiferno and Sandblast continue trap and damage opponents even if the user switches out.
- Multi-Attack uses the same power calculation as Fighting and Poison-type moves.
- When an active Dynamaxed Pokémon levels up, its max and current HP increase! This doesn't happen if a party Pokémon not in play levels up (lots of bugs have been squashed).
- G-Max Stun Shock and G-Max Befuddle choose their status conditions before considering immunities and can apply different status conditions to each opponent.

</details>

## What needs to be done? ##
- [x] RNG-based tests need to be rewritten!
- [x] When a player partner Pokemon is fainted by Max Flare, the game may softlock.
- Not in this PR, but in 2.0, Dynamax Level and Gigantamax factor need to be implemented.
- Not in this PR, but Gigantamax is not technically a form change, and should be reworked later to use some undecided swanky sprite system that Edu talked about. 

### DynamaxData struct ###
In the same vein of Mega Evolution and Z-Moves, a struct is added to `gBattleStruct` containing battle-related Dynamax data. Its fields and their purposes are listed below (note that numbers are used instead of constants for brevity):

- `bool8 playerSelect` - toggled every time the player presses START while Dynamax is available, checked on move selection to mark a battler as ready to Dynamax; still active after Dynamax, but does nothing.
- `u8 triggerSpriteId` - stores sprite ID for the Dynamax trigger
- `u8 indicatorSpriteId[4]` - stores sprite IDs for each battler's Dynamax indicator, since I wasn't able to find "room" in the healthbox `data` section
- `bool8 toDynamax[4]` - stores whether a battler has chosen to Dynamax
- `bool8 alreadyDynamaxed[2]` - stores whether a side has already Dynamaxed
- `bool8 dynamaxed[4]` - stores whether a battler is currently Dynamaxed; used for the function `IsDynamaxed(battler)`, which is used by most code for style.
- `u8 dynamaxTurns[4]` - stores how many turns of Dynamax a battler has remaining, decremented in `DoBattlerEndTurnEffects`
- `u8 usingMaxMove[4]` - stores whether a battler is using a Max Move, would be marginally redundant if not for the fact that Raid Bosses are not always using a Max Move despite being Dynamaxed; more details in the Dynamaxing section.
- `u8 activeSplit` - stores whether the current Max Move is physical or special
- `u8 splits[4]` - stores whether the base move a Dynamaxed battler selected is physical or special; both of these are modeled after Z-Moves
- `u16 baseMove[4]` - stores the base move used by a Dynamaxed battler; used for effects like Encore and Grudge
- `u16 lastUsedBaseMove` - stores the last used base move on the field; used for effects like Copycat (unfortunately just Copycat)
- `u16 levelUpHP` - stores the HP a Dynamaxed battler has before level-up to avoid having it reset

### Dynamaxing ###
The activation mechanism is copied straight from Mega Evolution. A bit flag is set when a move is chosen, and this is checked to mark a Pokémon as ready to Dynamax using `gBattleStruct->dynamax.toDynamax[battler]`. In `CheckMegaEvolutionBeforeTurn`, the Pokémon will have the appropriate fields filled using `PrepareBattlerForDynamax` and `BattleScript_DynamaxBegins` is executed. This script will perform the Dynamax animation (note the tint occurs in `battle_gfx_sfx_util`, just search `// dynamax tint`) and call the various command `updatedynamax`, which is also used when Dynamax ends. This command calls `RecalcBattlerStats` and updates the battler's healthbox.

The Dynamax HP Multiplier is handled through `ApplyDynamaxHPMultiplier`, which will make it very easy to handle Dynamax Level in the future. This function is called in `RecalcBattlerStats`, as well as on level-up to reapply it after HP is recalculated.

The trigger sprite and indicators are also copied directly from Mega Evolution, which is especially nice since the latter was recently completely reworked so there should be no bugs. The Dynamax indicator sprite shares a palette with the Alpha and Omega indicators, since this palette is always loaded and I want to avoid bloat.

### Max Move Usage ###
In `HandleTurnActionSelectionState`, a check is made using `ShouldUseMaxMove` to decide whether to mark the `usingMaxMove` field for a battler. Currently, this function just checks if a battler is Dynamaxed, but it will also check Raid Bosses in the future. Once marked, the selected move will be transformed in `HandleAction_UseMove`, which will also mark the correct move split to be used. If a status move is selected, its priority will be set to that of Max Guard in `GetMovePriority`. The field is cleared in `HandleAction_ActionFinished`.

Max Move effects are handled using a `SetMaxMoveEffect` script command and function in `battle_dynamax.c`. This command functions as a custom `setmoveeffect`, since Sheer Force and Serene Grace don't apply and I wanted to keep code self-contained instead of adding ~30 new move effect constants. Because of this, Max Moves have a custom script that is almost identical to `EffectHit`, however `tryfaintmon` is called before `setmaxmoveeffect`, which may be causing a bug?

### Max Move Scripts ###
Max Move effect scripts all follow a similar loop, storing the original target or attacker, trying to apply their effect, moving to the target's ally, then restoring the original target or attacker once the script ends. This was modeled after Jungle Healing's script, since I wanted to avoid looping through all battlers unnecessarily.

G-Max Steelsurge uses a script similar to Stone Axe (thanks Lunos!), and its hazards (generally just referred to in code as `steelsurge`) are handled just like Stealth Rocks.

G-Max Wildfire, Cannonade, Vine Lash, and Volcalith share a script and a side status called `damageNonTypes`. The timer and type unaffected by the damage-over-time effect are stored in `gSideTimers`.

The status-inducing effects are handled using two scripts (`trysetstatus1` and `trysetstatus2`) which are very generalized and simply check the current move's argument to decide which status to apply. This allows for easy randomization in the case of G-Max Befuddle and Stun Shock, and spares the expansion ~10 additional various commands (especially since Torment, Mean Look, Infatuation, and Confusion are compressed into one).

The various `jumpiftargetdynamaxed` is used to prevent the phazing effect of phazing moves. There is also a various for `tryhealsixthhealth` (which uses Dynamaxed HP) and `tryrecycleberry`, since G-Max Replenish only works for berries.

### Gigantamax ###
Gigantamax Forms are handled as species, with sprites, cries, and species info like everything else. Form-changing is handled using `FORM_CHANGE_BATTLE_GIGANTAMAX`, added to Edu's Form Change refactor. Everything is pretty straightforward, but it's worth noting that because form changing calls `RecalcBattlerStats`, the call to this function is avoided when Dynamaxing to avoid calling it twice and thus applying the Dynamax multiplier twice.

### Miscellaneous ###
Weight-based moves are checked and stopped in `attackcanceler`, modeled after how the effects of Primal Groudon and Kyogre's weather cancelling Water and Fire-type moves, respectively. A check is made using `IsMoveBlockedByDynamax` so that this might be expanded for Max Raids.

Max Guard's funky protect interactions are handled in `IsBattlerProtected` using a check to `IsMoveBlockedByMaxGuard`.

Max Move battlestrings print the full name of a Max Move similar to how Z-Moves are printed using a call to `GetMaxMoveName` and a table.

Almost every instance where `gBattleMoveDamage` was assigned a value based off the target's max HP or current HP have been replaced with an assignment using `GetNonDynamaxMaxHP` or `GetNonDynamaxHP`.

### Tests ###
Tests have been added for every Max Move effect, as well as many of the specific interactions listed a section or two above. G-Max Chi Strike is marked with `KNOWN_FAILING` since there seems to be an unrelated issue with moves with +3 crit stages not having 100% chance to be a critical hit. A couple other moves are currently marked with `KNOWN_FAILING` because they need to have their tests rewritten, but they are perfectly functional.

## **Discord Contact Info**
Agustin#1522

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
I will try to help as much as I can with this :)
That said, here are some pointers that I've noticed already:
- [ ] You added Alolan Form, Castform and Cherrim image duplicates. Please remove those.
- [ ] Use current folder format. `/gmax_alcremie/` -> `/alcremie/gmax/`
- [ ] Pleas take the chance to combine the Regis into Legendary too.
- [ ] Spaces instead of tabs.
- [ ] I recommend defining Max Moves/G-Max moves only once, and setting the move category based on the base move, like how Z-Moves are handled currently.
- [ ] Max Move power should have their own field in `gBattleMoves`, following this table: https://www.smogon.com/forums/threads/national-dex-mechanics-thread.3656786/post-8518210
- [ ] Please order `gFormChangeTablePointers` by base form's gen.
- [ ] `sMelmetalFormSpeciesIdTable` should go in Gen 7, not 8
- [ ] Remember to use `changedSpecies` to store the original form for species that gigantamax into the same form (Toxtricity and Alcremie). I'll be streamlining how it's handled in #1714 to help out.

**Comment by AgustinGDLV** [comment]
Max Move Showcase

Max Moves are mostly functional. This is what's left:
- [ ] Max Overgrowth doesn't set Grassy Terrain despite my frustration.
- [ ] Gigantamax Moves aren't able to be used or have coded effects.
- [ ] Custom battle strings for Max Moves that include the whole move name.
- [ ] No proper check on when to use Max Moves (currently always returns true since Dynamax doesn't exist).
- [ ] Max Moves bypass Protect entirely, as opposed to dealing 25% damage.
- [ ] Potential jank with how move split is decided.
- [ ] Max Move interactions with -ate abilities, Multi Attack, Weather Ball, Electrify, Terrain Pulse, Nature Power, etc.
- [ ] Convert `gDynamaxMovePowers` to a field in `gBattleMoves`.

**Comment by AgustinGDLV** [comment]
**Update:**
Every Max Move and G-Max Move is completed and functional! To consider them completely done, I want to write tests that check the effect of every Max Move, as well as edge cases and the like found in this [research thread](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/).

My TODO before Max Moves are _completely_ done is below. I'd love to have some of the changes reviewed and get some input on whether these things should be reworked.
- [ ] Write tests as mentioned above. 
- [ ] Rework the battle scripts to use `setallytonexttarget` instead of my custom `jumpiftargetnotally` to get rid of a `various`.
- [ ] Finish `GetMaxMove` by reworking `gMaxMovePowers` to only contain exceptions and become a static variable, as well as properly choose Gigantamax moves with a species check.
- [ ] Ensure Copycat, Encore, and Grudge function correctly against Max Moves, and Substitute doesn't prevent Max effects.
- [ ] Move animations or placeholder animations.

Almost there!

**Comment by AgustinGDLV** [comment]
Now marked as ready for review! Note, this requires and uses #2411. There's a little bit left (see checklist), but lots to look over.

**Comment by AgustinGDLV** [comment]
READY FOR REVIEW! There is one unconfirmed bug that I need to test and fix, however I just updated the hell out of this PR's description to make it easier for understand and comb through. I could use some help on rewriting a couple tests, but I may handle this myself soon.

File: test/dynamax.c [hunk 1]
```diff
@@ -0,0 +1,1386 @@
...
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_DRAGON_TAIL); MOVE(player, MOVE_TACKLE, dynamax: TRUE); }
+        TURN { MOVE(opponent, MOVE_WHIRLWIND); MOVE(player, MOVE_TACKLE); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Foe Wobbuffet used Dragon Tail!");
+        HP_BAR(player);
```
Comment (CallmeEchoo) [inline test/dynamax.c:136]: I think this will fail here. if I understand correctly, `HP_BAR(player)` fails if the hp bar of the player changes at all. 
If dragon tail is supposed to do damage, here this will fail. maybe @mrgriffin can confirm. 
if thats the case the other uses of `HP_BAR(player)` should be looked at aswell

File: test/dynamax.c [hunk 2]
```diff
@@ -0,0 +1,1386 @@
...
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_RED_CARD); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_CELEBRATE); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, opponent);
+        MESSAGE("Foe Wobbuffet held up its Red Card against Wobbuffet!");
+        MESSAGE("The move was blocked by the power of Dynamax!");
+    } FINALLY {
```
Comment (CallmeEchoo) [inline test/dynamax.c:158]:
```suggestion
    } THEN {
```
you know

File: test/dynamax.c [hunk 3]
```diff
@@ -0,0 +1,1386 @@
...
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_TACKLE); SEND_OUT(player, 1); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Foe Wobbuffet used Tackle!");
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
+        MESSAGE("Wobbuffet is switched out with the Eject Button!");
+    } FINALLY {
```
Comment (CallmeEchoo) [inline test/dynamax.c:176]:
```suggestion
    } THEN {
```

File: test/dynamax.c [hunk 4]
```diff
@@ -0,0 +1,1386 @@
...
+        ASSUME(P_GEN_8_POKEMON == TRUE);
+        PLAYER(SPECIES_MILTANK) { Ability(ABILITY_SCRAPPY); }
+        OPPONENT(SPECIES_RUNERIGUS) { Ability(ABILITY_WANDERING_SPIRIT); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_SKILL_SWAP); }
+    } SCENE {
+        MESSAGE("Miltank used Max Strike!");
+        MESSAGE("Foe Runerigus used Skill Swap!");
+        MESSAGE("But it failed!");
+    } FINALLY {
```
Comment (CallmeEchoo) [inline test/dynamax.c:193]:
```suggestion
    } THEN {
```

File: test/dynamax.c [hunk 5]
```diff
@@ -0,0 +1,1386 @@
...
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { Ability(ABILITY_SHADOW_TAG); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_SIMPLE_BEAM); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Foe Wobbuffet used Simple Beam!");
+        MESSAGE("Wobbuffet acquired Simple!");
+    } FINALLY {
```
Comment (CallmeEchoo) [inline test/dynamax.c:209]:
```suggestion
    } THEN {
```

File: test/dynamax.c [hunk 6]
```diff
@@ -0,0 +1,1386 @@
...
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_ENCORE); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Foe Wobbuffet used Encore!");
+        MESSAGE("But it failed!");
```
Comment (CallmeEchoo) [inline test/dynamax.c:224]: maybe use a different move after for extra proof

File: test/dynamax.c [hunk 7]
```diff
@@ -0,0 +1,1386 @@
...
+    }
+}
+
+SINGLE_BATTLE_TEST("(DYNAMAX) Dynamaxed Pokemon can be encored immediately after reverting")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { Speed(50); };
+        OPPONENT(SPECIES_WOBBUFFET) { Speed(100); };
```
Comment (CallmeEchoo) [inline test/dynamax.c:232]: I dont think speed needs to be explicitly stated here bc we can determine move order during the turn

File: test/dynamax.c [hunk 8]
```diff
@@ -0,0 +1,1386 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_ARM_THRUST, dynamax: TRUE); }
+        TURN { MOVE(player, MOVE_ARM_THRUST); }
+        TURN { MOVE(player, MOVE_ARM_THRUST); }
+        TURN { MOVE(opponent, MOVE_ENCORE); MOVE(player, MOVE_TACKLE); }
```
Comment (CallmeEchoo) [inline test/dynamax.c:237]: I think we need to use `FORCED_MOVE(player)` here instead of `MOVE(player, MOVE_TACKLE)`
see encore tests.
```
SINGLE_BATTLE_TEST("Encore forces consecutive move uses for 2 turns")
{
    GIVEN {
        PLAYER(SPECIES_WOBBUFFET);
        OPPONENT(SPECIES_WOBBUFFET);
    } WHEN {
        TURN { MOVE(player, MOVE_CELEBRATE); MOVE(opponent, MOVE_ENCORE); }
        TURN { FORCED_MOVE(player); }
        TURN { FORCED_MOVE(player); }
        TURN { MOVE(player, MOVE_SPLASH); }
    } SCENE {
        ANIMATION(ANIM_TYPE_MOVE, MOVE_CELEBRATE, player);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_ENCORE, opponent);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_CELEBRATE, player);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_CELEBRATE, player);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_SPLASH, player);
    }
}
```

File: test/dynamax.c [hunk 9]
```diff
@@ -0,0 +1,1386 @@
...
+        PLAYER(SPECIES_WOBBUFFET) { Item(ITEM_POTION); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_KNOCK_OFF); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Foe Wobbuffet used Knock Off!");
+        MESSAGE("Foe Wobbuffet knocked off Wobbuffet's Potion!");
+    }
+}
```
Comment (CallmeEchoo) [inline test/dynamax.c:310]: possibly
```
THEN {
   EXPECT_EQ(player->item, ITEM_NONE)
}
```

File: test/dynamax.c [hunk 10]
```diff
@@ -0,0 +1,1386 @@
...
+    bool32 protected;
+    KNOWN_FAILING; // Rounding issue?
+    PARAMETRIZE { protected = FALSE; }
+    PARAMETRIZE { protected = TRUE; }
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        if (protected)
+            TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_PROTECT); }
```
Comment (CallmeEchoo) [inline test/dynamax.c:373]: maybe double check that turn order is correct? without specified speed it is decided by which player acts first in `TURN`. not sure if priority overwrites.

File: test/dynamax.c [hunk 11]
```diff
@@ -0,0 +1,1386 @@
...
+// TODO: Gigantamax factor
+SINGLE_BATTLE_TEST("(DYNAMAX) Pokemon with Gigantamax forms change upon Dynamaxing")
+{
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR_GMAX);
```
Comment (CallmeEchoo) [inline test/dynamax.c:443]:
```suggestion
    } THEN {
        EXPECT_EQ(player->species, SPECIES_VENUSAUR_GMAX);
```

File: test/dynamax.c [hunk 12]
```diff
@@ -0,0 +1,1386 @@
...
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); }
+        TURN { SWITCH(player, 1); }
+        TURN { SWITCH(player, 0); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR);
```
Comment (CallmeEchoo) [inline test/dynamax.c:458]:
```suggestion
    } THEN {
        EXPECT_EQ(player->species, SPECIES_VENUSAUR);
```

File: test/dynamax.c [hunk 13]
```diff
@@ -0,0 +1,1386 @@
...
+
+SINGLE_BATTLE_TEST("(DYNAMAX) Pokemon with Gigantamax forms revert upon fainting")
+{
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR) { HP(1); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_TACKLE); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR);
```
Comment (CallmeEchoo) [inline test/dynamax.c:470]:
```suggestion
    } THEN {
        EXPECT_EQ(player->species, SPECIES_VENUSAUR);
```
can we even check the player form at this point?

File: test/dynamax.c [hunk 14]
```diff
@@ -0,0 +1,1386 @@
...
+    }
+}
+
+// Move selection tests can't be simulated.
```
Comment (CallmeEchoo) [inline test/dynamax.c:474]: what do you mean by this? maybe you need this? 
```
 * FORCED_MOVE(battler)
 * Used when the battler chooses Fight and then their move is chosen for
 * them, e.g. when affected by Encore.
 *     FORCED_MOVE(player);
```

**Review by CallmeEchoo — COMMENTED** [review]
I didnt get to the end unfortunately. but I think this might be a good start :)

File: test/dynamax.c (hunk 1 above)
Comment (AgustinGDLV) [reply test/dynamax.c:136]: I thought `HP_BAR` was supposed to confirm that there was a change in health. That's how I've been treating it.

File: test/dynamax.c (hunk 10 above)
Comment (AgustinGDLV) [reply test/dynamax.c:373]: I'm 90% sure priority overwrites.

File: test/dynamax.c (hunk 1 above)
Comment (mrgriffin) [reply test/dynamax.c:150]: Your understanding is correct, I made a mistake when writing the documentation for `HP_BAR`, which has since been corrected :)
